tests/components/nobo_hub/test_config_flow.py

```diff
@@ -0,0 +1,362 @@
+            "device": 0,
+            "serial_suffix": "",
+        },
+    )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "missing_serial_suffix"}
+
+
+async def test_configure_invalid_serial_manual(hass: HomeAssistant) -> None:
+    """Test we handle invalid serial error."""
+    with patch(
+        "pynobo.nobo.async_discover_hubs",
+        return_value=[],
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": "manual"}
+        )
+
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {"ip_address": "1.1.1.1", "serial": "123456789"},
+    )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "invalid_serial"}
+
+
+async def test_configure_invalid_ip_address(hass: HomeAssistant) -> None:
+    """Test we handle invalid ip address error."""
+    with patch(
+        "pynobo.nobo.async_discover_hubs",
+        return_value=[("1.1.1.1", "123456789")],
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": "manual"}
+        )
+
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {"serial": "123456789012", "ip_address": "ABCD"},
+    )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "invalid_ip"}
+
+
+async def test_configure_cannot_connect(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    with patch(
+        "pynobo.nobo.async_discover_hubs",
+        return_value=[("1.1.1.1", "123456789")],
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_USER}
+        )
+
+    with patch(
+        "pynobo.nobo.async_connect_hub",
+        return_value=False,
+    ) as mock_connect:
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {"device": 0, "serial_suffix": "012"},
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "cannot_connect"}
+    mock_connect.assert_awaited_once_with("1.1.1.1", "123456789012")
+
+
+async def test_options_flow(hass: HomeAssistant) -> None:
+    """Test the options flow."""
+    zones = {
+        "1": {"zone_id": "1", "name": "Kitchen"},
+        "2": {"zone_id": "2", "name": "Bedrooms"},
+    }
+    week_profiles = {
+        "1": {"week_profile_id": "1", "name": "Off"},
+        "2": {"week_profile_id": "2", "name": "Kitchen On"},
+        "3": {"week_profile_id": "3", "name": "Bedrooms On"},
+    }
+    hub = Mock(nobo)
+    type(hub).zones = PropertyMock(return_value=zones)
+    type(hub).week_profiles = PropertyMock(return_value=week_profiles)
+    entry = MockConfigEntry(
+        domain="nobo_hub",
+        unique_id="123456789012",
+        data={"serial": "123456789012"},
+    )
+    hass.data[DOMAIN] = {
```

Comment:
Don't interact with integration details, like hass.data, in tests. Set up the integration while patching the library to avoid I/O. Then start the options flow.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

Note that we may remove the options flow for select entities.

Reply:
Thanks! Fixed.